src/mdns/mdns_mdnssd.cpp

```diff
@@ -552,19 +665,33 @@ void PublisherMDnsSd::HandleRegisterHostResult(DNSServiceRef       aHostsConnect
     OTBR_UNUSED_VARIABLE(aHostsConnection);
     OTBR_UNUSED_VARIABLE(aFlags);
 
-    auto host =
-        std::find_if(mHosts.begin(), mHosts.end(), [&](const Host &aHost) { return aHost.mRecord == aHostRecord; });
-    assert(host != mHosts.end());
+    HostIterator host = FindPublishedHost(aHostRecord);
+    std::string  hostName;
+
+    VerifyOrExit(host != mHosts.end());
+    hostName = host->mName;
+
+    otbrLog(OTBR_LOG_INFO, "[mdns] got a reply for host %s", host->mName);
 
     if (aErrorCode == kDNSServiceErr_NoError)
     {
-        otbrLog(OTBR_LOG_INFO, "Successfully registered host %s", host->mName);
+        otbrLog(OTBR_LOG_INFO, "[mdns] successfully registered host %s", host->mName);
     }
     else
     {
-        otbrLog(OTBR_LOG_WARNING, "Failed to register host %s for mdnssd error: %s", host->mName,
+        otbrLog(OTBR_LOG_WARNING, "[mdns] failed to register host %s for mdnssd error: %s", host->mName,
                 DNSErrorToString(aErrorCode));
+
+        DiscardHost(host->mName, /* aSendGoodbye */ false);
+    }
+
+    if (mHostHandler != nullptr)
+    {
+        mHostHandler(host->mName, DNSErrorToOtbrError(aErrorCode), mHostHandlerContext);
```

Comment:
should use `hostName` here in case host is discarded. 
```suggestion
        mHostHandler(hostName, DNSErrorToOtbrError(aErrorCode), mHostHandlerContext);
```

Reply:
Thanks for catching this! Yes this is what the intention of this variable but somehow forgot to update the usage.